server/src/main/java/org/elasticsearch/common/text/UTF8DecodingReader.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.common.text;
+
+import org.elasticsearch.xcontent.XContentString;
+
+import java.io.Reader;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Reader that decodes UTF-8 formatted bytes into chars.
+ */
+public class UTF8DecodingReader extends Reader {
```

Comment:
This is basically equivalent to:
```java
final var reader = new InputStreamReader(new ByteArrayInputStream(utfBytes.bytes(), utfBytes.offset(), utfBytes.length()), StandardCharsets.UTF_8);
```

But according to the microbenchmarks, using the `InputStreamReader/ByteArrayInputStream` is really slow, and was actually slower than the original string-based implementation.

Reply:
Here are the results from the microbenchmark:
```
Benchmark                                                  (nDocs)  Mode  Cnt    Score   Error  Units
OptimizedTextBenchmark.indexDocuments (baseline)           1048576  avgt    5  581.242 ± 3.050  ms/op
OptimizedTextBenchmark.indexDocuments (UTF8DecodingReader) 1048576  avgt    5  544.477 ± 4.961  ms/op
OptimizedTextBenchmark.indexDocuments (InputStreamReader)  1048576  avgt    5  852.380 ± 6.238  ms/op
```

Reply:
I haven't yet, but I plan to today

Reply:
Maybe make this a final class?